File: 85. Maximal Rectangle/85. Maximal Rectangle.dart
class Solution {
  int maximalRectangle(List<List<String>> matrix) {
    if (matrix.isEmpty || matrix[0].isEmpty) return 0;
    int rows = matrix.length, cols = matrix[0].length;
    List<int> heights = List.filled(cols, 0);
    int maxArea = 0;

    for (int i = 0; i < rows; i++) {
      for (int j = 0; j < cols; j++) {
        if (matrix[i][j] == '1') {
          heights[j]++;
        } else {
          heights[j] = 0;
        }
      }
      maxArea = max(maxArea, largestRectangleArea(heights));
    }

    return maxArea;
  }

  int largestRectangleArea(List<int> heights) {
    List<int> stack = [];
    int maxArea = 0;
    int i = 0;

    while (i <= heights.length) {
      int h = (i == heights.length) ? 0 : heights[i];
      if (stack.isEmpty || h >= heights[stack.last]) {
        stack.add(i);
        i++;
      } else {
        int height = heights[stack.removeLast()];
        int width = stack.isEmpty ? i : i - stack.last - 1;
        maxArea = max(maxArea, height * width);
      }
    }

    return maxArea;
  }

  int max(int a, int b) => a > b ? a : b;
}

File: 85. Maximal Rectangle/85. Maximal Rectangle.rs
impl Solution {
    pub fn maximal_rectangle(matrix: Vec<Vec<char>>) -> i32 {
        if matrix.is_empty() || matrix[0].is_empty() {
            return 0;
        }
        let (rows, cols) = (matrix.len(), matrix[0].len());
        let mut heights = vec![0; cols];
        let mut max_area = 0;

        for i in 0..rows {
            for j in 0..cols {
                if matrix[i][j] == '1' {
                    heights[j] += 1;
                } else {
                    heights[j] = 0;
                }
            }
            max_area = max_area.max(Self::largest_rectangle_area(&heights));
        }

        max_area
    }

    fn largest_rectangle_area(heights: &[i32]) -> i32 {
        let mut stack = Vec::new();
        let mut max_area = 0;
        let mut heights = heights.to_vec();
        heights.push(0);

        for (i, &h) in heights.iter().enumerate() {
            while !stack.is_empty() && h < heights[*stack.last().unwrap()] {
                let height = heights[stack.pop().unwrap()];
                let width = if stack.is_empty() {
                    i as i32
                } else {
                    (i - stack.last().unwrap() - 1) as i32
                };
                max_area = max_area.max(height * width);
            }
            stack.push(i);
        }

        max_area
    }
}